core/templates/dev/head/services/RteHelperService.js

```diff
@@ -181,9 +181,26 @@ oppia.factory('RteHelperService', [
           });
         });
 
-        text = elt.html();
-        text = $filter('removeExtraLines')(text);
-        return text;
+        var textElt = elt[0].childNodes;
+        for (var i = textElt.length - 1; i > 0; i--) {
+          for (var j = textElt[i].childNodes.length; j > 0; j--) {
+            if (textElt[i].childNodes[j - 1].nodeValue === null ||
+              textElt[i].childNodes[j - 1].nodeValue.trim() === '') {
+              textElt[i].childNodes[j - 1].remove();
+            }
```

Comment:
Suppose you have something along the lines of the following structure:

`<p> text <br> text <br> text <br> </p>`.

Is the intended result `<p> text <br> text <br> text </p>` or `<p> text text text </p>`? If the former then this code might not be doing what you want it to do.

Also, let's say the paragraph contains `<img>` tags, or other tags which have a null value. Do these get removed as well? I don't understand why we're checking for null nodes rather than e.g. just `<br>` and `<p>`.

Cases like these should be reflected in the tests.

Reply:
> <p> text <br> text <br> text <br> </p>

Yes that's a good fix! I missed that.

>I don't understand why we're checking for null nodes rather than e.g. just <br> and <p>

Because of paragraph containing only spaces. I made a new fix, please check if it's fine